Testnet/NFTAuctionStore/NFTAuctionStore/NFTAuctionStore.cs

```diff
@@ -0,0 +1,228 @@
+{
+    private void SetAuctionInfo(Address contract, UInt256 tokenId, AuctionInfo auctionInfo)
+    {
+        State.SetStruct($"AuctionInfo:{contract}:{tokenId}", auctionInfo);
+    }
+
+    public AuctionInfo GetAuctionInfo(Address contract, UInt256 tokenId)
+    {
+        return State.GetStruct<AuctionInfo>($"AuctionInfo:{contract}:{tokenId}");
+    }
+
+    public ulong GetRefund(Address address)
+    {
+        return State.GetUInt64($"Refund:{address}");
+    }
+
+    private void SetRefund(Address address, ulong balance)
+    {
+        State.SetUInt64($"Refund:{address}", balance);
+    }
+
+    public NFTAuctionStore(ISmartContractState state) : base(state)
+    {
+        EnsureNotPayable();
+    }
+
+    public void Bid(Address contract, UInt256 tokenId)
+    {
+        var auction = GetAuctionInfo(contract, tokenId);
+
+        Assert(auction.OnAuction, "Auction is not found.");
+
+        Assert(!EndBlockReached(auction), "Auction ended.");
+
+        Assert(Message.Value > auction.HighestBid && Message.Value >= auction.StartingPrice, "The amount is not higher than highest bidder or starting price.");
+
+        if (auction.HighestBid > 0)
+        {
+            //refund for previous bidder
+            var balance = GetRefund(auction.HighestBidder);
+
+            SetRefund(auction.HighestBidder, balance + auction.HighestBid);
+        }
+
+        auction.HighestBidder = Message.Sender;
+        auction.HighestBid = Message.Value;
+
+        SetAuctionInfo(contract, tokenId, auction);
+
+        Log(new HighestBidUpdatedLog { Contract = contract, TokenId = tokenId, Bidder = auction.HighestBidder, Bid = auction.HighestBid });
+    }
+
+    private bool EndBlockReached(AuctionInfo auction)
+    {
+        return Block.Number >= auction.EndBlock;
+    }
+
+    public bool Refund()
+    {
+        EnsureNotPayable();
+
+        var amount = GetRefund(Message.Sender);
+
+        Assert(amount > 0);
+
+        SetRefund(Message.Sender, 0);
+
+        var transfer = Transfer(this.Message.Sender, amount);
+
+        Assert(transfer.Success, "Transfer failed.");
+
+        Log(new BalanceRefundedLog { To = Message.Sender, Amount = amount });
+
+        return transfer.Success;
+    }
+
+    public void AuctionEnd(Address contract, UInt256 tokenId)
+    {
+        EnsureNotPayable();
+
+        var auction = GetAuctionInfo(contract, tokenId);
+
+        Assert(auction.OnAuction, "Auction is not found.");
+
+        Assert(EndBlockReached(auction), "Auction is not ended yet.");
+
+        auction.OnAuction = false;
+        SetAuctionInfo(contract, tokenId, auction);
+
+        if (auction.HighestBid == 0)
+        {
+            TransferToken(contract, tokenId, Address, auction.Seller);
+
+            Log(new AuctionEndFailedLog { Contract = contract, TokenId = tokenId });
+            return;
+        }
+
+        var result = Transfer(auction.Seller, auction.HighestBid);
```

Comment:
These funds should instead be a withdrawable balance for the seller.

Reply:
I made improvement in here. I refund when seller is a contract